pkg/kubefnord/join.go

```diff
@@ -0,0 +1,678 @@
+		},
+		Spec: crv1a1.ClusterSpec{
+			KubernetesAPIEndpoints: crv1a1.KubernetesAPIEndpoints{
+				ServerEndpoints: []crv1a1.ServerAddressByClientCIDR{
+					{
+						ClientCIDR:    "0.0.0.0/0",
+						ServerAddress: host,
+					},
+				},
+			},
+		},
+	}
+
+	if dryRun {
+		return cluster, nil
+	}
+
+	cluster, err := crClientset.ClusterregistryV1alpha1().Clusters().Create(cluster)
+	if err != nil {
+		return cluster, err
+	}
+
+	return cluster, nil
+}
+
+// getClientsets sets up the clientsets necessary for the join operations.
+func getClientsets(hostConfig, clusterConfig, fedConfig *rest.Config) (util.FedClientset, error) {
+	fedClientset := util.NewFedClientset()
+
+	_, err := fedClientset.NewHostClientset(hostConfig)
+	if err != nil {
+		glog.V(2).Infof("Failed to get host cluster clientset: %v", err)
+		return nil, err
+	}
+
+	_, err = fedClientset.NewClusterClientset(clusterConfig)
+	if err != nil {
+		glog.V(2).Infof("Failed to get joining cluster clientset: %v", err)
+		return nil, err
+	}
+
+	_, err = fedClientset.NewFedClientset(fedConfig)
+	if err != nil {
+		glog.V(2).Infof("Failed to get federation clientset: %v", err)
+		return nil, err
+	}
+
+	return fedClientset, nil
+}
+
+// createFederatedCluster creates a federated cluster resource that associates
+// the cluster and secret.
+func createFederatedCluster(fedClientset *fedclient.Clientset, joiningClusterName,
+	secretName string, dryRun bool) (*fedv1a1.FederatedCluster, error) {
+	fedCluster := &fedv1a1.FederatedCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: joiningClusterName,
+		},
+		Spec: fedv1a1.FederatedClusterSpec{
+			ClusterRef: corev1.LocalObjectReference{
+				Name: joiningClusterName,
+			},
+			SecretRef: &corev1.LocalObjectReference{
+				Name: secretName,
+			},
+		},
+	}
+
+	if dryRun {
+		return fedCluster, nil
+	}
+
+	fedCluster, err := fedClientset.FederationV1alpha1().FederatedClusters().Create(fedCluster)
+
+	if err != nil {
+		return fedCluster, err
+	}
+
+	return fedCluster, nil
+}
+
+// getCMDeployment retrieves the deployment in the federation namespace and
+// checks if it matches the name of the controller manager and if so, returns
+// it.
+func getCMDeployment(clientset client.Interface,
+	fedNamespace string) (*appsv1.Deployment, error) {
+	depList, err := clientset.AppsV1().Deployments(fedNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, dep := range depList.Items {
+		if strings.HasSuffix(dep.Name, CMNameSuffix) {
+			return &dep, nil
+		}
+	}
+	return nil, fmt.Errorf("could not find the deployment for federation controller manager in host cluster")
+}
+
+// getFederationName gets the federation name from the federation
+// deployment.
+func getFederationName(hostClientSet client.Interface, fedNamespace string) (string, error) {
+	d, err := getCMDeployment(hostClientSet, fedNamespace)
+	if err != nil {
+		// TODO: Figure out how to handle this for integration tests.
+		return "", err
```

Comment:
This will need to change when we support running `JoinCluster` from the integration tests because we do not have a deployment running the federation controller/apiserver. Once we get a k8s namespace controller in integration tests we can look at using `JoinCluster` and revisit this then.